<a href="https://colab.research.google.com/github/ChronosXV/Aprendizaje-Automatico/blob/main/Aprendizaje_Automatico.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**CONSUMO DIGITAL EN ADOLESCENTES**

Analisis exploratorio

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
import pandas as pd
df = pd.read_excel("/content/drive/MyDrive/Aprendizaje automatico/repositorio.xlsx")

In [6]:
# Estado inicial del dataset
print("Filas:", df.shape[0])
print("Columnas:", df.shape[1])

Filas: 4115
Columnas: 125


Visualizamos el tipo de dato de cada columna, si tenemos filas duplicadas, columnas con el mismo nombre repetido

In [10]:
print(df.dtypes.value_counts())
print(df.columns[df.columns.duplicated()]) # nombres de columna repetidos

float64           66
int64             37
object            21
datetime64[ns]     1
Name: count, dtype: int64
Index([], dtype='object')


No tenemos duplicados

In [ ]:
# vemos las distintas medidas
df.describe()

,Submission ID,Submission Date,investigador,Colegio,Ciudad,grado,edad,sexonum,Altura en m,Peso en kg,...,matematica,prom,aplazos,datoscompletoDEMOG,datoscompletoSUENO,datoscompletoPANTALLAS,excSUENOpantallas,excluidos,missingnotas,excluFINAL
count,4.115000e+03,4115,4115.000000,4115.000000,4115.000000,4115.000000,3553.000000,4115.000000,4081.000000,4056.000000,...,1257.000000,1257.000000,1257.000000,4115.000000,4115.000000,4115.000000,4115.000000,4115.000000,4115.000000,4115.000000
mean,4.407193e+18,2019-08-04 07:13:31.672903936,9.879708,43.297448,18.820656,3.215796,15.600338,0.446659,1.653658,57.782544,...,7.175123,7.308385,0.430390,0.863426,0.782503,0.869502,0.305225,0.404860,0.694532,0.790036
min,4.347980e+18,2019-05-27 18:26:20,1.000000,1.000000,0.000000,1.000000,12.000000,0.000000,1.400000,27.000000,...,1.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,4.373565e+18,2019-06-26 09:08:15,3.000000,29.000000,14.000000,2.000000,14.000000,0.000000,1.600000,50.000000,...,6.000000,6.500000,0.000000,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,1.000000
50%,4.422135e+18,2019-08-21 14:17:34,7.000000,47.000000,16.000000,3.000000,16.000000,0.000000,1.650000,56.000000,...,7.000000,7.500000,0.000000,1.000000,1.000000,1.000000,0.000000,0.000000,1.000000,1.000000
75%,4.435872e+18,2019-09-06 11:53:09,15.000000,62.000000,27.000000,4.000000,17.000000,1.000000,1.720000,65.000000,...,9.000000,8.500000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000
max,4.494709e+18,2019-11-13 13:15:30,53.000000,89.000000,92.000000,6.000000,20.000000,1.000000,1.910000,90.000000,...,10.000000,10.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000
std,3.442841e+16,NaN,8.537678,22.669166,11.285266,1.518031,1.601527,0.497207,0.092749,11.320989,...,1.804322,1.440699,0.495328,0.343438,0.412594,0.336892,0.460559,0.490925,0.460661,0.407332


### Exploración de valores faltantes

Vamos a identificar las columnas con valores faltantes y el porcentaje de datos que faltan en cada una.

In [ ]:
# Buscamos valores nulos
print("Valores nulos por columna:")
print(df.isnull().sum())

Valores nulos por columna:
IP                        0
Submission ID             0
Submission Date           0
Código                    0
investigador              0
                         ..
datoscompletoPANTALLAS    0
excSUENOpantallas         0
excluidos                 0
missingnotas              0
excluFINAL                0
Length: 125, dtype: int64


In [ ]:
# Buscamos duplicados
cantidad_duplicados = df.duplicated().sum()

print(f"Cantidad de filas duplicadas: {cantidad_duplicados}")

Cantidad de filas duplicadas: 0


### Búsqueda de duplicados en columnas clave

Investigamos si existen duplicados basándonos en un subconjunto de columnas que deberían identificar entradas únicas, como 'IP' y 'Submission ID'.

In [ ]:
key_columns = ['IP', 'Submission ID']
duplicates_in_keys = df[df.duplicated(subset=key_columns, keep=False)]

print(f"Cantidad de filas duplicadas en '{', '.join(key_columns)}': {len(duplicates_in_keys)}")

if not duplicates_in_keys.empty:
    print("Primeras filas duplicadas:")
    display(duplicates_in_keys.sort_values(by=key_columns).head())
else:
    print(f"No se encontraron duplicados en las columnas {', '.join(key_columns)}.")

Cantidad de filas duplicadas en 'IP, Submission ID': 0
No se encontraron duplicados en las columnas IP, Submission ID.


### Revisión de valores faltantes

Cantidad de valores faltantes por columna

In [11]:
# % de faltantes por columna, ordenado de mayor a menor
faltantes = df.isna().mean().sort_values(ascending=False) * 100
print(faltantes.head(30).round(2))

print()
# Los indicadores de completitud que ya trae el dataset
for col in ['datoscompletoDEMOG', 'datoscompletoSUENO', 'datoscompletoPANTALLAS', 'excluidos', 'excluFINAL']:
    if col in df.columns:
        print(col, '->', df[col].value_counts(dropna=False).to_dict())

WAZ                                              99.34
prom                                             69.45
aplazos                                          69.45
matematica                                       69.45
lengua                                           69.36
tpotvonlinePORC                                  27.85
tpotabletPORC                                    27.85
tporedesPORC                                     27.85
tpovideojuegosPORC                               27.85
tpootrasPORC                                     27.85
tpopantallatotalhstrimmed                        27.70
tpopantallatotalhscat                            27.70
tpopantallatotalhscat2                           27.70
BAZ                                              22.26
BMIzscoreCAT                                     22.26
HAZ                                              22.19
tposueñosemTRIMMED                               17.11
tposueñosemCAT                                   17.11
edad      

De esta forma detectamos que la columna WAZ tiene un porcentaje muy alto de valores faltantes. Por otro lado, las columnas prom, aplazos, matematica, lengua (~69% faltante) — el rendimiento académico está mayormente vacío

In [12]:
flags = ['datoscompletoDEMOG', 'datoscompletoSUENO', 'datoscompletoPANTALLAS', 'excluFINAL']

# Cuántas filas quedan si exijo cada combinación
print('Solo excluFINAL==1:', (df['excluFINAL'] == 1).sum())
print('Solo pantallas completas:', (df['datoscompletoPANTALLAS'] == 1).sum())
print('Pantallas + sueño completos:', ((df['datoscompletoPANTALLAS'] == 1) & (df['datoscompletoSUENO'] == 1)).sum())
print('Los 3 bloques completos:', ((df['datoscompletoDEMOG'] == 1) & (df['datoscompletoSUENO'] == 1) & (df['datoscompletoPANTALLAS'] == 1)).sum())
print('Los 3 bloques + excluFINAL:', ((df['datoscompletoDEMOG'] == 1) & (df['datoscompletoSUENO'] == 1) & (df['datoscompletoPANTALLAS'] == 1) & (df['excluFINAL'] == 1)).sum())

# Edad: ¿el 13.66% de faltantes está dentro o fuera de excluFINAL==1?
print()
print('Edad faltante dentro de excluFINAL==1:', df.loc[df['excluFINAL'] == 1, 'edad'].isna().sum())
print('Edad faltante fuera de excluFINAL==1:', df.loc[df['excluFINAL'] == 0, 'edad'].isna().sum())

Solo excluFINAL==1: 3251
Solo pantallas completas: 3578
Pantallas + sueño completos: 2859
Los 3 bloques completos: 2449
Los 3 bloques + excluFINAL: 1748

Edad faltante dentro de excluFINAL==1: 399
Edad faltante fuera de excluFINAL==1: 163


Distintos criterios de filtro

In [14]:
flags = ['datoscompletoDEMOG', 'datoscompletoSUENO', 'datoscompletoPANTALLAS', 'excluFINAL']

# Cuántas filas quedan si exijo cada combinación
print('Solo excluFINAL==1:', (df['excluFINAL'] == 1).sum())
print('Solo pantallas completas:', (df['datoscompletoPANTALLAS'] == 1).sum())
print('Pantallas + sueño completos:', ((df['datoscompletoPANTALLAS'] == 1) & (df['datoscompletoSUENO'] == 1)).sum())
print('Los 3 bloques completos:', ((df['datoscompletoDEMOG'] == 1) & (df['datoscompletoSUENO'] == 1) & (df['datoscompletoPANTALLAS'] == 1)).sum())
print('Los 3 bloques + excluFINAL:', ((df['datoscompletoDEMOG'] == 1) & (df['datoscompletoSUENO'] == 1) & (df['datoscompletoPANTALLAS'] == 1) & (df['excluFINAL'] == 1)).sum())

# Edad: ¿el 13.66% de faltantes está dentro o fuera de excluFINAL==1?
print()
print('Edad faltante dentro de excluFINAL==1:', df.loc[df['excluFINAL'] == 1, 'edad'].isna().sum())
print('Edad faltante fuera de excluFINAL==1:', df.loc[df['excluFINAL'] == 0, 'edad'].isna().sum())

Solo excluFINAL==1: 3251
Solo pantallas completas: 3578
Pantallas + sueño completos: 2859
Los 3 bloques completos: 2449
Los 3 bloques + excluFINAL: 1748

Edad faltante dentro de excluFINAL==1: 399
Edad faltante fuera de excluFINAL==1: 163


**Hallazgo importante**: dentro de excluFINAL == 1 (3.251 filas), igual falta edad en 399 casos (12,3%) — casi la misma proporción que fuera de ese filtro (163 de 864 = 18,9%). Eso quiere decir que excluFINAL no garantiza completitud demográfica. El flag de los autores se construyó mirando solo sueño y/o pantallas, no las variables demográficas que necesitamos  para caracterizar los clusters. Por lo tanto  hay que agregar el propio filtro de completitud demográfica.